sematic/scheduling/job_scheduler.py

```diff
@@ -0,0 +1,158 @@
+
+    Parameters
+    ----------
+    run:
+        The run to schedule
+    resolution:
+        The resolution associated with the run
+    """
+    run.external_jobs = _refresh_external_jobs(run.external_jobs)
+    _validate_scheduleable(run, resolution)
+    external_jobs_list = list(run.external_jobs) + [_schedule_job(run, resolution)]
+    run.external_jobs = tuple(external_jobs_list)
+    run.future_state = FutureState.SCHEDULED
+    return run
+
+
+def update_run_status(
+    future_state: FutureState,
+    external_jobs: Union[List[ExternalJob], Tuple[ExternalJob, ...]],
+) -> Tuple[FutureState, Optional[str], Tuple[ExternalJob, ...]]:
+    """Determine whether a new run state should be used based ONLY external job statuses
+
+    The external jobs themselves will have their state information refreshed before
+    determining whether the run needs its status changed.
+
+    Parameters
+    ----------
+    future_state:
+        The current state of the run
+    external_jobs:
+        The external jobs associated with the run.
+
+    Returns
+    -------
+    A tuple with 3 elements. The first is the new future state (same state if unchanged).
+    The second is an optional message for why the state changed. The third is new external
+    jobs, updated.
+    """
+    external_jobs = tuple(external_jobs)
+    if future_state.is_terminal():
+        return future_state, None, external_jobs
+    if future_state.value == FutureState.RAN.value:
+        # If the job already RAN, the only reason it's not
+        # terminal is because child runs have to complete.
+        # There should be no more external jobs for this run.
+        return future_state, None, external_jobs
+    if future_state.value == FutureState.CREATED.value:
+        if len(external_jobs) == 0:
+            return future_state, None, external_jobs
+        else:
+            raise ValueError(
+                "Run is in an invalid state: it is marked as CREATED but it has "
+                "external jobs. Runs with external jobs should be SCHEDULED."
+            )
+    if len(external_jobs) < 1:
+        raise ValueError("No external jobs for run")
+    external_jobs = _refresh_external_jobs(external_jobs)
+    if future_state.value == FutureState.SCHEDULED.value:
+        if not any(job.is_active() for job in external_jobs):
+            job_summary_str = "; ".join([repr(job) for job in external_jobs])
+            logger.warning(
+                "Job failed due to K8s job failure. Job states: %s", job_summary_str
+            )
+            return (
+                FutureState.FAILED,
+                "The kubernetes job(s) experienced an unknown failure",
+                external_jobs,
+            )
+        return FutureState.SCHEDULED, None, external_jobs
+    raise ValueError(
+        f"Future is in a state not covered by update logic: {future_state}"
+    )
+
+
+def _validate_scheduleable(run: Run, resolution: Resolution):
```

Comment:
Naming nit: `_assert_is_scheduleable` or something. Makes it clear it will raise if invalid.

Reply:
done